networkx/algorithms/community/demon.py

```diff
@@ -0,0 +1,296 @@
+# -*- coding: utf-8 -*-
+#    Copyright (C) 2015
+#    All rights reserved.
+#    BSD license.
+
+"""
+Flat Merge version of DEMON (Democratic Estimate of the Modular Organization of
+a Network) algorithm as described in:
+
+Michele Coscia, Giulio Rossetti, Fosca Giannotti, Dino Pedreschi:
+DEMON: a local-first discovery method for overlapping communities.
+KDD 2012:615-623
+
+DEMON uses a local-first approach to community discovery were each node
+democratically votes for the communities that it sees in its surroundings. It
+provides a deterministic, fully incremental method for community discovery that
+has a limited time complexity.
+
+Based on the implementation by Giulio Rossetti <giulio.rossetti@isti.cnr.it>:
+http://www.michelecoscia.com/wp-content/uploads/2013/07/demon_py.zip
+"""
+
+from __future__ import division
+import networkx as nx
+import random
+
+__all__ = ["demon_communities"]
+
+
+def demon_communities(G, epsilon=0.25, min_community_size=3, weight=None):
+    """Finds the communities in a graph using the label Democratic Estimate of
+    the Modular Organization of a Network (DEMON) algorithm[1]_. This method
+    uses a local-first approach to community discovery were each node
+    democratically votes for the communities that it sees in its surroundings.
+    Communities in this algorithm can be overlapping
+
+    Algorithm:
+    ----------
+    Require: G(V, E), C' = 0, epsilon = [0..1]
+    Ensure: set of overlapping communities C'
+
+    1: for all v in V do:
+    2:   e <- EgoMinusEgo(v, G)
+    3:   C(v) <- LabelPropagation(e)
+    4:   for all C in C'(v) do:
+    5:     C <- C union v
+    6:     C' <- Merge(C', C, v)
+    7:   end for
+    8: end for
+    9: return C'
+
+    Parameters
+    ----------
+    G : graph
+        An NetworkX graph.
+
+    epsilon: real
+        The tolerance required in order to merge communities. The epsilon
+        factor is introduced to vary the percentage of common elements provided
+        from each couple of communities: epsilon = 0 ensure that two
+        communities are merged only if one of them is a proper subset of the
+        other, on the other hand with a value of epsilon = 1 even communities
+        that do not share a single node are merged together
+
+    min_community_size: integer
+        Min nodes needed to form a community
+
+    weight: String
+        If the graph is weighted, the edge parameter representing the weight
+
+    Returns
+    -------
+    communities : list
+        List of overlapping communities (sets)
+
+    References
+    ----------
+    .. [1] Michele Coscia, Giulio Rossetti, Fosca Giannotti, Dino Pedreschi:
+           DEMON: a local-first discovery method for overlapping communities.
+           KDD 2012:615-623.
+    """
+
+    for n in G.nodes():
+            G.node[n]['communities'] = [n]
```

Comment:
You are modifying the input graph, possibly overwriting user-defined node attributes. Why is this necessary? If it really is necessary, it should be clearly documented.